### DISCUSSION TITLE
Fail to Complie after update EmbUI to V3.0.0

### DISCUSSION THREAD

**Discussion by Alaa-Aim** [vortigont/espem#39] [General]
@vortigont , after updating to EmbUI V3.0.0, I get the error as below :

espem/interface.cpp: In function 'void create_parameters()':
espem/interface.cpp:46:11: error: 'class EmbUI' has no member named 'section_handle_add'
     embui.section_handle_add(B_ESPEM, block_page_main);              // generate "main" info page
           ^~~~~~~~~~~~~~~~~~
espem/interface.cpp:47:11: error: 'class EmbUI' has no member named 'section_handle_add'
     embui.section_handle_add(B_ESPEMSET, block_page_espemset);       // generate "ESPEM settings" page
           ^~~~~~~~~~~~~~~~~~
espem/interface.cpp:53:14: error: 'add_sections' is not a member of 'basicui'
     basicui::add_sections();
              ^~~~~~~~~~~~
espem/interface.cpp:57:11: error: 'class EmbUI' has no member named 'section_handle_add'
     embui.section_handle_add(A_SET_ESPEM,  set_sampler_opts);
           ^~~~~~~~~~~~~~~~~~
espem/interface.cpp:58:11: error: 'class EmbUI' has no member named 'section_handle_add'
     embui.section_handle_add(A_SET_UART,   set_uart_opts);
           ^~~~~~~~~~~~~~~~~~
espem/interface.cpp:59:11: error: 'class EmbUI' has no member named 'section_handle_add'
     embui.section_handle_add(A_SET_PZOPTS,   set_pzopts);
           ^~~~~~~~~~~~~~~~~~
espem/interface.cpp:62:11: error: 'class EmbUI' has no member named 'section_handle_add'
     embui.section_handle_add(A_DIRECT_CTL,  set_directctrls);             // process direct update controls
           ^~~~~~~~~~~~~~~~~~
espem/interface.cpp: In function 'void section_main_frame(Interface*, ArduinoJson::V6213PB2::JsonObject*)':
espem/interface.cpp:87:82: error: no matching function for call to 'Interface::json_section_manifest(const char* const&, int, const char [6])'
     interf->json_section_manifest(C_DICT[lang][CD::ESPEM_H], 0, FW_VERSION_STRING);    // HEADLINE for WebUI
                                                                                  ^
In file included from .pio/libdeps/espem/EmbUI/EmbUI/EmbUI.h:15,
                 from espem/interface.cpp:2:
.pio/libdeps/espem/EmbUI/EmbUI/ui.h:555:9: note: candidate: 'template<class ID, class L> typename std::enable_if<is_string_v<TString>, void>::type Interface::json_section_manifest(ID, const char*, unsigned int, L)'
         json_section_manifest(const ID appname, const char* devid, unsigned appjsapi = 0, const L appversion = P_EMPTY);
         ^~~~~~~~~~~~~~~~~~~~~
.pio/libdeps/espem/EmbUI/EmbUI/ui.h:555:9: note:   template argument deduction/substitution failed:
In file included from C:/Users/ALAA/.platformio/packages/framework-arduinoespressif32/cores/esp32/Arduino.h:30,
                 from .pio/libdeps/espem/EmbUI/EmbUI/globals.h:9,
                 from espem/main.h:31,
                 from espem/espem.h:10,
                 from espem/interface.cpp:1:
espem/interface.cpp:87:65: note:   cannot convert '"3.1.0"' (type 'const char [6]') to type 'unsigned int'
     interf->json_section_manifest(C_DICT[lang][CD::ESPEM_H], 0, FW_VERSION_STRING);    // HEADLINE for WebUI
                                                                 ^
espem/main.h:21:29: note: in expansion of macro 'TOSTRING'
 #define FW_VERSION_STRING   TOSTRING(FW_VERSION_MAJOR) "." TOSTRING(FW_VERSION_MINOR) "." TOSTRING(FW_VERSION_REVISION)
                             ^~~~~~~~
espem/interface.cpp:87:65: note: in expansion of macro 'FW_VERSION_STRING'
     interf->json_section_manifest(C_DICT[lang][CD::ESPEM_H], 0, FW_VERSION_STRING);    // HEADLINE for WebUI
                                                                 ^~~~~~~~~~~~~~~~~
espem/interface.cpp:95:18: error: 'block_settings_netw' is not a member of 'basicui'
         basicui::block_settings_netw(interf, data);
                  ^~~~~~~~~~~~~~~~~~~
espem/interface.cpp:95:18: note: suggested alternative: 'page_settings_netw'
         basicui::block_settings_netw(interf, data);
                  ^~~~~~~~~~~~~~~~~~~
                  page_settings_netw
espem/interface.cpp: In function 'void block_menu(Interface*, ArduinoJson::V6213PB2::JsonObject*)':
espem/interface.cpp:116:14: error: 'opt_setup' is not a member of 'basicui'
     basicui::opt_setup(interf, data);       // пункт меню "настройки"
              ^~~~~~~~~
Archiving .pio\build\espem\lib75b\libAsyncTCP.a
espem/espem.cpp: In member function 'void ESPEM::wspublish()':
espem/espem.cpp:233:49: error: no matching function for call to 'Interface::Interface(EmbUI*, AsyncWebSocket*, int)'
   Interface interf(&embui, &embui.ws, PUB_JSSIZE);
                                                 ^
In file included from .pio/libdeps/espem/EmbUI/EmbUI/EmbUI.h:15,
                 from espem/espem.cpp:10:
.pio/libdeps/espem/EmbUI/EmbUI/ui.h:433:9: note: candidate: 'Interface::Interface(AsyncWebServerRequest*, size_t)'
         Interface(AsyncWebServerRequest *request, size_t size = IFACE_DYN_JSON_SIZE): json(size), _delete_handler_on_destruct(true) {
         ^~~~~~~~~
.pio/libdeps/espem/EmbUI/EmbUI/ui.h:433:9: note:   candidate expects 2 arguments, 3 provided
.pio/libdeps/espem/EmbUI/EmbUI/ui.h:430:9: note: candidate: 'Interface::Interface(AsyncWebSocketClient*, size_t)'
         Interface(AsyncWebSocketClient *client, size_t size = IFACE_DYN_JSON_SIZE): json(size), _delete_handler_on_destruct(true) {
         ^~~~~~~~~
.pio/libdeps/espem/EmbUI/EmbUI/ui.h:430:9: note:   candidate expects 2 arguments, 3 provided
.pio/libdeps/espem/EmbUI/EmbUI/ui.h:427:9: note: candidate: 'Interface::Interface(AsyncWebSocket*, size_t)'
         Interface(AsyncWebSocket *server, size_t size = IFACE_DYN_JSON_SIZE): json(size), _delete_handler_on_destruct(true) {
         ^~~~~~~~~
.pio/libdeps/espem/EmbUI/EmbUI/ui.h:427:9: note:   candidate expects 2 arguments, 3 provided
.pio/libdeps/espem/EmbUI/EmbUI/ui.h:423:9: note: candidate: 'Interface::Interface(FrameSend*, size_t)'
         Interface (FrameSend *feeder, size_t size = IFACE_DYN_JSON_SIZE): json(size), _delete_handler_on_destruct(false) {
         ^~~~~~~~~
.pio/libdeps/espem/EmbUI/EmbUI/ui.h:423:9: note:   candidate expects 2 arguments, 3 provided
.pio/libdeps/espem/EmbUI/EmbUI/ui.h:368:7: note: candidate: 'Interface::Interface(const Interface&)'
 class Interface {
       ^~~~~~~~~
.pio/libdeps/espem/EmbUI/EmbUI/ui.h:368:7: note:   candidate expects 1 argument, 3 provided
Compiling .pio\build\espem\lib5f0\WiFi\WiFiClient.cpp.o
Compiling .pio\build\espem\lib5f0\WiFi\WiFiGeneric.cpp.o
Archiving .pio\build\espem\lib0ad\libFS.a
*** [.pio\build\espem\src\interface.cpp.o] Error 1
*** [.pio\build\espem\src\espem.cpp.o] Error 1
==================================================================== [FAILED] Took 21.70 seconds ====================================================================
Environment    Status    Duration
-------------  --------  ------------
espem          FAILED    00:00:21.703

**Comment by vortigont** [comment]
Hello @Alaa-Aim,
pls do not try to change EmbUI version for this project except one that is defined in `platformio.ini`
For this project it must be `https://github.com/vortigont/EmbUI.git#v2.8.1`.
EmbUI ver 3.0.0 has significant API changes that is not yet  implemented in **espem**. I will update it when time permits.
It will be a major change and introduce full MQTT support for the project

**Reply by Alaa-Aim** [reply]
@vortigont many thanks, waiting for your coming project update...
